### ISSUE TITLE
Development Mode Doesn't Include Modules Conditionally Required by Instrumentation Since 14.2

### ISSUE THREAD

**Issue by CHC383** [vercel/next.js#64802]
### Link to the code that reproduces this issue

https://github.com/CHC383/nextjs-nextlogger-config-bug-example/tree/dev-mode-module-bug

### To Reproduce

Start the application in dev mode (`npm run dev`) and check the console output

### Current vs. Expected behavior

- Current behavior: `Error: An error occurred while loading instrumentation hook: Cannot find module '/Users/xxx/workspaces/next-logger-bug-example/next-logger.config.cjs'`
- Expected behavior: The application could start and load the next-logger config

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.4.0: Fri Mar 15 00:10:42 PDT 2024; root:xnu-10063.101.17~1/RELEASE_ARM64_T6000
  Available memory (MB): 32768
  Available CPU cores: 8
Binaries:
  Node: 21.6.2
  npm: 10.2.4
  Yarn: 1.22.21
  pnpm: 8.14.3
Relevant Packages:
  next: 14.3.0-canary.12 // Latest available version is detected (14.3.0-canary.12).
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.1.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Instrumentation, Module Resolution

### Which stage(s) are affected? (Select all that apply)

next dev (local)

### Additional context

[Next-Logger](https://github.com/sainsburys-tech/next-logger) is a patcher to allow using Pino for Next internal logging, and it uses [Cosmiconfig](https://github.com/cosmiconfig/cosmiconfig) to load the config located in the root of the project, which internally uses [import-fresh](https://github.com/sindresorhus/import-fresh).

The example works fine until 14.2.0, last known working version is 14.1.4. And after 14.2.0, it still works in our production setup, which runs the app inside docker and has the next-logger config in the working dir, the only affecting mode I found so far is the dev mode.

Similar report in the next-logger repo: https://github.com/sainsburys-tech/next-logger/issues/29

Full error log:
```
❯ npm run dev  

> dev
> next dev

  ▲ Next.js 14.3.0-canary.12
  - Local:        http://localhost:3000
  - Experiments (use with caution):
    · instrumentationHook

 ✓ Starting...
 ⚠ ./node_modules/cosmiconfig/dist/loaders.js
Critical dependency: the request of a dependency is an expression

Import trace for requested module:
./node_modules/cosmiconfig/dist/loaders.js
./node_modules/cosmiconfig/dist/index.js
./node_modules/next-logger/lib/logger.js
./node_modules/next-logger/lib/patches/next.js
./node_modules/next-logger/presets/all.js
./node_modules/next-logger/index.js
Error: An error occurred while loading instrumentation hook: Cannot find module '/Users/xxxx/workspaces/next-logger-bug-example/next-logger.config.cjs'
    at webpackEmptyContext (/Users/xxxx/workspaces/next-logger-bug-example/.next/server/instrumentation.js:44:10)
    at module.exports (webpack-internal:///(instrument)/./node_modules/import-fresh/index.js:32:110)
    at loadJsSync (webpack-internal:///(instrument)/./node_modules/cosmiconfig/dist/loaders.js:17:12)
    at #loadConfiguration (webpack-internal:///(instrument)/./node_modules/cosmiconfig/dist/ExplorerSync.js:91:24)
    at #readConfiguration (webpack-internal:///(instrument)/./node_modules/cosmiconfig/dist/ExplorerSync.js:77:74)
    at search (webpack-internal:///(instrument)/./node_modules/cosmiconfig/dist/ExplorerSync.js:44:63)
    at emplace (webpack-internal:///(instrument)/./node_modules/cosmiconfig/dist/util.js:12:20)
    at ExplorerSync.search (webpack-internal:///(instrument)/./node_modules/cosmiconfig/dist/ExplorerSync.js:71:42)
    at eval (webpack-internal:///(instrument)/./node_modules/next-logger/lib/logger.js:8:30)
    at (instrument)/./node_modules/next-logger/lib/logger.js (/Users/xxxx/workspaces/next-logger-bug-example/.next/server/vendor-chunks/next-logger.js:39:1)
    at __webpack_require__ (/Users/xxxx/workspaces/next-logger-bug-example/.next/server/webpack-runtime.js:33:43)
    at eval (webpack-internal:///(instrument)/./node_modules/next-logger/lib/patches/next.js:3:16)
    at (instrument)/./node_modules/next-logger/lib/patches/next.js (/Users/xxxx/workspaces/next-logger-bug-example/.next/server/vendor-chunks/next-logger.js:59:1)
    at __webpack_require__ (/Users/xxxx/workspaces/next-logger-bug-example/.next/server/webpack-runtime.js:33:43)
    at eval (webpack-internal:///(instrument)/./node_modules/next-logger/presets/all.js:1:1)
    at (instrument)/./node_modules/next-logger/presets/all.js (/Users/xxxx/workspaces/next-logger-bug-example/.next/server/vendor-chunks/next-logger.js:69:1)
    at __webpack_require__ (/Users/xxxx/workspaces/next-logger-bug-example/.next/server/webpack-runtime.js:33:43)
    at eval (webpack-internal:///(instrument)/./node_modules/next-logger/index.js:1:1)
    at (instrument)/./node_modules/next-logger/index.js (/Users/xxxx/workspaces/next-logger-bug-example/.next/server/vendor-chunks/next-logger.js:19:1)
    at __webpack_require__ (/Users/xxxx/workspaces/next-logger-bug-example/.next/server/webpack-runtime.js:33:43)
    at Module.register (webpack-internal:///(instrument)/./instrumentation.ts:9:15)
    at async DevServer.runInstrumentationHookIfAvailable (/Users/xxxx/workspaces/next-logger-bug-example/node_modules/next/dist/server/dev/next-dev-server.js:437:17)
    at async Span.traceAsyncFn (/Users/xxxx/workspaces/next-logger-bug-example/node_modules/next/dist/trace/trace.js:154:20)
    at async DevServer.prepareImpl (/Users/xxxx/workspaces/next-logger-bug-example/node_modules/next/dist/server/dev/next-dev-server.js:214:9)
    at async NextServer.prepare (/Users/xxxx/workspaces/next-logger-bug-example/node_modules/next/dist/server/next.js:161:13)
    at async initializeImpl (/Users/xxxx/workspaces/next-logger-bug-example/node_modules/next/dist/server/lib/render-server.js:98:5)
    at async initialize (/Users/xxxx/workspaces/next-logger-bug-example/node_modules/next/dist/server/lib/router-server.js:424:22)
    at async Server.<anonymous> (/Users/xxxx/workspaces/next-logger-bug-example/node_modules/next/dist/server/lib/start-server.js:250:36) {
  code: 'MODULE_NOT_FOUND',
  filepath: '/Users/xxxx/workspaces/next-logger-bug-example/next-logger.config.cjs'
}
```

**Comment by CHC383** [comment]
Looks like a Webpack module detection problem, and needs extra config to make `next-logger` work in the development mode, see https://github.com/sainsburys-tech/next-logger/issues/13#issuecomment-2113625702 for the solution

**Comment by CHC383** [comment]
See https://github.com/sainsburys-tech/next-logger/pull/34 for the details of webpack `require`, this has been fixed on `next-logger` since 5.0.0
